### PR TITLE
feat(platform): record A/B experiment arms per user

### PR DESCRIPTION
**Pull request by ntindle**
### Why / What / How

**Why:** Experiment arms need a durable per-user record so activation, retention, and cost can be compared with the variant the user first saw. The pricing experiment also needs to wait for its variant before applying a billing default.

**What:** This adds authenticated experiment assignment recording and frontend hooks for PostHog and LaunchDarkly string variants. It depends on #14286.

**How:** `ExperimentAssignment` keeps the first assignment for each user and experiment. The hooks report resolved string variants through the generated API client. LaunchDarkly exposures are deduplicated by user, flag, and variant, and a disabled PostHog integration does not consume an exposure. Disabled feature flags produce no variant, assignment, or exposure even when the SDK retains a string value. Transient assignment failures retry at most twice with backoff. Cleanup cancels pending retries and releases unfinished claims; token ownership prevents an old request's late failure from erasing a newer claim.

### Changes 🏗️

- Add the experiment assignment migration, authenticated GET/POST endpoints, data access, and API schema.
- Add `useExperiment`, `useLaunchDarklyExperiment`, and shared assignment reporting; wait for resolution before applying the pricing default.
- Cover same-browser user changes, disabled analytics, successful and exhausted retries, StrictMode, unmount/account changes, and in-flight request replacement.
- No environment variables or Compose changes. The migration applies with the normal backend migration process.

`data/*.py` authorization: the authenticated router passes the JWT user ID into both recording and listing. Clients cannot supply another user's ID; the API tests assert that user-scoped argument.

### Agents and large language models used

- Claude Code — model unknown (original implementation).
- Codex — GPT-6 (review fixes, parent integration, and validation).

#### For code changes:

- [x] I have clearly listed my changes in the PR description.
- [x] I have made a test plan.
- [x] I have tested my changes according to the test plan:
  - [x] 40 experiment-hook and pricing integration tests pass on parent `fea7f594`, using the generated API client, RTL, and MSW. Five original regressions and one additional in-flight lifecycle regression were reproduced before their fixes.
  - [x] 4 mock-backed API tests pass using the existing isolated unit fixture harness. No live database migration test was run locally.
  - [x] Whole frontend format/lint/types and backend/library format/Pyright pass on the integrated parent.
  - [x] The broader pre-integration frontend run completed with 6,682 passing tests and four failures in the unrelated tour end-card file. All four passed on one isolated rerun with network permissions; no tour source changes were made. The complete suite was not repeated after parent integration.
  - [x] Normal pre-commit and pre-push hooks.

#### For configuration changes:

- [x] `.env.default` is already compatible with these changes.
- [x] `docker-compose.yml` is already compatible with these changes.
- [x] Configuration changes are listed above.

---

> [!NOTE]
> **Medium Risk**
> Introduces new persisted user data and changes onboarding pricing timing when PostHog loads; incorrect assignment recording would skew experiment analytics but endpoints are auth-scoped and variants are immutable after first write.
> 
> **Overview**
> Adds **durable per-user experiment assignments** so server-side analytics can join activation, retention, and cost to the variant a user first saw, independent of PostHog/LaunchDarkly history.
> 
> **Backend:** New `ExperimentAssignment` table (unique per user + experiment key), migration, and data layer with **first-write-wins** semantics (later re-buckets are logged but not overwritten). Authenticated **`GET`/`POST` `/api/experiments/assignments`** records/list assignments scoped to the JWT user; OpenAPI is updated accordingly.
> 
> **Frontend:** Shared **`useReportAssignment`** posts resolved string variants once per signed-in user (with backoff retries and claim/dedup logic). **`useExperiment`** wraps PostHog multivariate flags with **`isResolved`**; **`useLaunchDarklyExperiment`** waits for LD init, reports to the same API, and fires a one-time PostHog `experiment_exposed` event. Onboarding **subscription pricing** now waits for flag resolution before applying the default billing cycle to avoid UI flash and wrong defaults.
> 
> **Tests:** API unit tests plus hook tests for enrollment, deduplication, account switches, disabled analytics/flags, and retry/unmount edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6b3a195af7e229951814768c327045369f864a5a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: autogpt_platform/backend/migrations/20260902120100_add_experiment_assignment/migration.sql
Comment (Pwuts) [inline autogpt_platform/backend/migrations/20260902120100_add_experiment_assignment/migration.sql:1]: Table creation without a schema.prisma change??

File: autogpt_platform/backend/backend/api/rest_api.py [hunk 1]
```diff
@@ -391,6 +392,11 @@ async def validation_error_handler(
     prefix="/api/analytics",
     tags=["analytics"],
 )
+app.include_router(
+    backend.api.features.experiments.router,
+    prefix="/api/experiments",
+    tags=["v2", "experiments"],
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/api/rest_api.py]:
```suggestion
    tags=["experiments"],
```

File: autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  const flags = useFlags<Record<string, unknown>>();
+  const client = useLDClient();
+  const [isClientReady, setIsClientReady] = useState(false);
+  const flagsEnabled = environment.areFeatureFlagsEnabled();
+
+  useEffect(() => {
+    if (!client) return;
+    let active = true;
+    client
+      .waitForInitialization(LD_READY_TIMEOUT_MS)
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts]: Good catch, `waitForInitialization` takes seconds. Fixed in d302ac9 (5 seconds, matching the onboarding page).

File: autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  const { mutate: recordAssignment } = usePostV2RecordExperimentAssignment();
+  const userID = user?.id ?? null;
+
+  useEffect(() => {
+    if (!isResolved || !variant || !userID) return;
+    if (!claimAssignmentReport(userID, experimentKey)) return;
+    recordAssignment({
+      data: { experiment_key: experimentKey, variant, source },
+    });
+  }, [isResolved, variant, userID, experimentKey, source, recordAssignment]);
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts:44]: Fixed in d302ac9: a failed report releases the per-session claim in `onError`, so the next render retries against the idempotent endpoint.

File: autogpt_platform/backend/migrations/20260902120100_add_experiment_assignment/migration.sql
Comment (ntindle) [reply autogpt_platform/backend/migrations/20260902120100_add_experiment_assignment/migration.sql:1]: The model had leaked into the base PR's schema (#14286) while the migration sat here. d302ac9 moves the `ExperimentAssignment` model into this PR next to its migration, and #14286 no longer declares it.

File: autogpt_platform/backend/backend/api/rest_api.py (hunk 1 above)
Comment (ntindle) [reply autogpt_platform/backend/backend/api/rest_api.py]: Applied in d302ac9, along with the renamed generated hook that follows from it.

File: autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+"use client";
+
+import { usePostExperimentsRecordExperimentAssignment } from "@/app/api/__generated__/endpoints/experiments/experiments";
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts:3]: False positive. Spaces in operationIds are this repo's existing convention (e.g. `getV2Export block cost estimates` on dev) and orval derives the hook name from them; `usePostExperimentsRecordExperimentAssignment` is the generated export and the experiments tests exercise it.

File: autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    experimentKey: flagKey,
+    variant,
+    isResolved,
+    source: "launchdarkly",
+  });
+
+  useEffect(() => {
+    if (!isResolved || !variant || !claimExposure(flagKey, variant)) return;
+    if (!environment.isPostHogEnabled()) return;
+    try {
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts]: Fixed in [f4bdb43ef](https://github.com/Significant-Gravitas/AutoGPT/commit/f4bdb43ef3c9a9fd4caf8f261730e97237847403): exposure deduplication now includes the authenticated user ID, flag, and variant. Signed-out users do not claim an exposure, and PostHog being disabled is checked before claiming. Regression tests cover switching between two users in the same mounted hook, signing in after flag resolution, and enabling PostHog without losing the exposure.

File: autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }, [isResolved, variant, flagKey]);
+
+  return { variant, isResolved };
+}
+
+const exposures = new Set<string>();
+
+function claimExposure(flagKey: string, variant: string) {
+  const key = `${flagKey}:${variant}`;
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts]: Fixed in [f4bdb43ef](https://github.com/Significant-Gravitas/AutoGPT/commit/f4bdb43ef3c9a9fd4caf8f261730e97237847403): the exposure effect waits for an authenticated user, includes user ID in its dependencies, and scopes claims by user/flag/variant. The disabled-PostHog guard runs before claiming. MSW/RTL coverage verifies both users receive their exposure while duplicate mounts for one user remain deduplicated.

File: autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  useEffect(() => {
+    if (!isResolved || !variant || !userID) return;
+    if (!claimAssignmentReport(userID, experimentKey)) return;
+    recordAssignment(
+      { data: { experiment_key: experimentKey, variant, source } },
+      {
+        // The backend is idempotent, so a failed report may be retried on
+        // the next render instead of being lost for the rest of the session.
+        onError: () => releaseAssignmentReport(userID, experimentKey),
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useReportAssignment.ts]: Fixed in [f4bdb43ef](https://github.com/Significant-Gravitas/AutoGPT/commit/f4bdb43ef3c9a9fd4caf8f261730e97237847403): transient assignment failures now retry at most twice with backoff through the generated mutation, without requiring changed arguments or remounting. Cleanup cancels scheduled work and releases unfinished claims; ownership tokens prevent a late failure from deleting a replacement claim. Regression tests use the real generated client with MSW and StrictMode, including first-failure/success, retry exhaustion, rejected input, account changes, unmount, and an old in-flight failure after remount.

File: autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        if (active) setIsClientReady(true);
+      });
+    return () => {
+      active = false;
+    };
+  }, [client]);
+
+  const isResolved = !flagsEnabled || isClientReady;
+  const raw = flags[flagKey];
+  const variant = typeof raw === "string" ? raw : null;
```
[the remark this answers is not in the document]
Comment (ntindle) [reply autogpt_platform/frontend/src/services/experiments/useLaunchDarklyExperiment.ts]: Fixed in [8a83aa8b3](https://github.com/Significant-Gravitas/AutoGPT/commit/8a83aa8b34aa4e8b6c4fe0c8ad2867fa9ed4ae98): variant derivation now requires feature flags to be enabled, so a retained SDK string cannot reach assignment reporting or exposure capture when disabled. The expanded regression first reproduced both an assignment POST and a PostHog exposure, then passed with the gate; it also checks immediate resolution with a null variant. All 40 experiment/pricing tests, whole frontend format/lint/types, and normal hooks pass.
